include/glow/Optimizer/FunctionPass.h

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_OPTIMIZER_FUNCTIONPASS_H
+#define GLOW_OPTIMIZER_FUNCTIONPASS_H
+
+namespace glow {
+
+class FunctionPass {
+protected:
+  /// Perform some tasks before the pass.
+  virtual void doInitialization(Function *F) const {
+    LOG_SCOPE(F->getLogContext(), getName());
+  }
+
+  /// Perform some tasks after the pass.
+  virtual void doFinalization(Function *F) const {}
+
+  /// Implementation of the pass. \returns whether the pass modifies \p F.
+  virtual bool passImpl(Function *F) = 0;
+
+public:
+  FunctionPass() = default;
+  virtual ~FunctionPass() = default;
+
+  // Run the pass, including initialization and finalization. \returns whether
+  // the pass modifies \p F.
+  virtual bool runOnFunction(Function *F) {
+    doInitialization(F);
```

Comment:
Do you expect each pass to perform pass-specific `doInitialization` and `doFinalization`? If the logic will be rather uniform for all passes, it should better be handled in the pass manager and then these two functions should not be a part of the API.

Reply:
I was thinking that there would be uniform logic for all passes handled in the pass manager. The flow when we run a pass would be:

```cpp
PassManager::prePass();
FunctionPass::runOnFunction(); // this calls:
//  FunctionPass::doInitialization();
//  FunctionPass::passImpl();
//  FunctionPass::doFinalization();
PassManager::postPass();
```

For now I put `LOG_SCOPE()` inside `doInitialization()` because we don't have the pass manager yet, but it would be moved into `PassManager::prePass()` eventually.

Reply:
Based on this, it seems like you'd be able to move virtually all of the logic from doInitialization and doFinalization into the pass manager (it will call LOG_SCOPE, etc for all passes it runs), which means that these APIs are probably not strictly necessary. 

I'm not against having doInitialization and doFinalization. I just think that they will be NOPs most of the time.

Reply:
Yeah agreed, I will remove `doInitialization` and `doFinalization` and let the pass implement its logic however it wants.